src/NumericalAlgorithms/LinearSolver/ConjugateGradient/InitializeElement.hpp

```diff
@@ -34,55 +35,89 @@ struct InitializeResidual;
-      &b, &Ax
-    ](const gsl::not_null<db::item_type<operand_tag>*> p) noexcept {
-      *p = b - Ax;
-    });
-    auto r = db::item_type<residual_tag>(get<operand_tag>(box));
+  template <typename DataBox, typename... InboxTags, typename ArrayIndex,
+            typename ActionList, typename ParallelComponent>
+  static auto apply(DataBox& box,
+                    const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    const Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const ArrayIndex& array_index, const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    db::mutate<operand_tag>(
+        make_not_null(&box),
+        [](const gsl::not_null<db::item_type<operand_tag>*> operand,
+           const db::item_type<source_tag>& source,
+           const db::item_type<operator_applied_to_fields_tag>&
+               operator_applied_to_fields) noexcept {
+          *operand = source - operator_applied_to_fields;
+        },
+        get<source_tag>(box), get<operator_applied_to_fields_tag>(box));
+    auto residual = db::item_type<residual_tag>{get<operand_tag>(box)};
 
     // Perform global reduction to compute initial residual magnitude square for
     // residual monitor
     Parallel::contribute_to_reduction<
         cg_detail::InitializeResidual<ParallelComponent>>(
         Parallel::ReductionData<
             Parallel::ReductionDatum<double, funcl::Plus<>>>{
-            inner_product(r, r)},
+            inner_product(residual, residual)},
         Parallel::get_parallel_component<ParallelComponent>(cache)[array_index],
         Parallel::get_parallel_component<ResidualMonitor<Metavariables>>(
             cache));
 
-    return db::create_from<db::RemoveTags<>, simple_tags, compute_tags>(
-        std::move(box),
-        // We have not started iterating yet, so the _next_ iteration ID is
-        // zero. We initialize the current iteration id such that it advances to
-        // zero by the increment operator.
-        std::numeric_limits<size_t>::max(), size_t{0}, std::move(r),
-        db::item_type<LinearSolver::Tags::HasConverged>{});
+    using compute_tags = db::AddComputeTags<
+        ::Tags::NextCompute<LinearSolver::Tags::IterationId>>;
+    return std::make_tuple(
+        ::Initialization::merge_into_databox<
+            InitializeElement,
+            db::AddSimpleTags<LinearSolver::Tags::IterationId, residual_tag,
+                              LinearSolver::Tags::HasConverged>,
+            compute_tags, MergePolicy>(
+            std::move(box),
+            // We have not started iterating yet, so we initialize the current
+            // iteration ID such that the _next_ iteration ID is zero.
+            std::numeric_limits<size_t>::max(), std::move(residual),
+            db::item_type<LinearSolver::Tags::HasConverged>{}),
+        // Terminate algorithm for now. The reduction will be broadcast to the
```

Comment:
I'm not certain it is the most intuitive strategy to terminate the algorithm here. Is there a good reason to prefer terminating here over including a separate `TerminatePhase` action in the action list or over using an `is_ready` to determine whether the reduction has been called?

Reply:
This is how global synchronizations are currently handled in the linear solver, see e.g. `ConjugateGradient/ElementActions.hpp`. I'm happy to discuss alternatives. However, that's a (potential) follow-up PR since it's a more involved change of how the linear solver algorithm works. To your suggestions:
- The approach with the `TerminatePhase` action wouldn't work I think. It's possible that the reduction comes in before the `TerminatePhase` action is reached in the action list, in which case the algorithm wouldn't be restarted.
- `is_ready` would require an additional action like, which the user has to remember to add to the action list. I like that this would - in principle - allow doing work while the reduction is processing. But during initialization this hardly matters and during the actual algorithm we need the result of the reduction to proceed with the linear solver step. So I don't currently see the additional complexity for the user compensated by a gain in performance. And going forward the idea is to improve parallelization not by avoiding or optimizing these global synchronization points, but by preconditioning the linear solver in a parallelizable way (i.e. Multigrid with Schwarz smoother).

Reply:
I see, thanks. I won't request a code change on this, then. My only concern is that it may not be obvious to users of this code creating an initialization action list that the initialization will not proceed after this action until the reduction occurs. If you agree that this is a misunderstanding to avoid, I'll leave it up to you where it's best to make a documentation note -- I'm not sure whether it should be at the start of this detail struct or preceding the typedefs which alias to template specializations of this struct.

Reply:
Such a note already exists for `prepare_step` in `ConjugateGradient.hpp` and `Gmres.hpp`, so I added it also to `initialize_element`. Please check the fixup :)